# **Feature Engineering**

Importacion de librerias


In [ ]:
!pip install -q geopandas shapely pyproj rtree

import pandas as pd
import numpy as np
import geopandas as gpd
from shapely.geometry import Point
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

Carga de los datasets


In [ ]:
import pandas as pd


In [ ]:
df_ventas = pd.read_csv('https://raw.githubusercontent.com/kiaranatale/Trabajo-Practico-Real-State-Analytics/refs/heads/main/Datos2/dataset_ventas_limpio.csv')

In [ ]:
df_ventas.head()

,Precio,Expensas,Calle,Altura,Link,Ambientes,Dormitorios,Baños,Antiguedad,Disposicion,...,fuente,_dup_interfuente,rango_sup,Moneda,Precio_num,Precio_USD,Precio_ARS,TC_CCL_aplicado,TC_fecha,ratio_expensas_precio
0,USD 2.500.000,2400000.0,Jerónimo Salguero,2700.0,https://www.argenprop.com/departamento-en-vent...,5.0,3.0,4,17.0,Frente,...,argenprop,False,"(103.0, 992.0]",USD,2500000.0,2500000.0,3.703000e+09,1481.2,2026-05-13T14:00:00.000Z,0.000648
1,USD 150.000,260000.0,Bulnes,1600.0,https://www.argenprop.com/departamento-en-vent...,3.0,2.0,1,30.0,Frente,...,argenprop,False,"(53.0, 72.0]",USD,150000.0,150000.0,2.221800e+08,1481.2,2026-05-13T14:00:00.000Z,0.001170
2,USD 113.000,116000.0,Ángel Justianiano Carranza,2300.0,https://www.argenprop.com/departamento-en-vent...,1.0,0.0,1,30.0,Frente,...,argenprop,False,"(39.0, 53.0]",USD,113000.0,113000.0,1.673756e+08,1481.2,2026-05-13T14:00:00.000Z,0.000693
3,USD 330.000,203300.0,ARAOZ,1200.0,https://www.argenprop.com/departamento-en-vent...,4.0,3.0,1,40.0,Contrafrente,...,argenprop,True,"(72.0, 103.0]",USD,330000.0,330000.0,4.887960e+08,1481.2,2026-05-13T14:00:00.000Z,0.000416
4,USD 270.000,300000.0,Honduras,3900.0,https://www.argenprop.com/departamento-en-vent...,4.0,2.0,2,20.0,Frente,...,argenprop,False,"(72.0, 103.0]",USD,270000.0,270000.0,3.999240e+08,1481.2,2026-05-13T14:00:00.000Z,0.000750


In [ ]:
df_alquileres=pd.read_csv('https://raw.githubusercontent.com/kiaranatale/Trabajo-Practico-Real-State-Analytics/refs/heads/main/Datos2/dataset_alquiler_limpio.csv')

In [ ]:
df_alquileres.head()

,Precio,Expensas,Calle,Altura,Link,Ambientes,Dormitorios,Baños,Antiguedad,Disposicion,...,fuente,_dup_interfuente,rango_sup,Moneda,Precio_num,Precio_USD,Precio_ARS,TC_CCL_aplicado,TC_fecha,ratio_expensas_precio
0,USD 950,20000.0,Manuel Nicolas Savio,400.0,https://www.argenprop.com/departamento-en-alqu...,3.0,2.0,2,35.0,Lateral,...,argenprop,False,"(52.274, 76.0]",USD,950.0,950.000000,1407140.0,1481.2,2026-05-13T14:00:00.000Z,0.147818
1,ARS 890.000,7633.0,Niceto vega,5900.0,https://www.argenprop.com/departamento-en-alqu...,1.0,0.0,1,35.0,Frente,...,argenprop,False,"(42.0, 52.274]",ARS,890000.0,600.864164,890000.0,1481.2,2026-05-13T14:00:00.000Z,0.196067
2,ARS 1.600.000,100000.0,El Salvador,5500.0,https://www.argenprop.com/departamento-en-alqu...,3.0,2.0,2,15.0,Contrafrente,...,argenprop,False,"(76.0, 900.0]",ARS,1600000.0,1080.205239,1600000.0,1481.2,2026-05-13T14:00:00.000Z,0.228125
3,USD 1.400,100000.0,RODRIGUEZ PEÑA,2000.0,https://www.argenprop.com/departamento-en-alqu...,5.0,2.0,3,40.0,Frente,...,argenprop,False,"(76.0, 900.0]",USD,1400.0,1400.000000,2073680.0,1481.2,2026-05-13T14:00:00.000Z,0.129721
4,ARS 900.000,25000.0,HONDURAS,4600.0,https://www.argenprop.com/departamento-en-alqu...,1.0,0.0,1,35.0,Contrafrente,...,argenprop,False,"(35.0, 42.0]",ARS,900000.0,607.615447,900000.0,1481.2,2026-05-13T14:00:00.000Z,0.155556


In [ ]:
df_calles = pd.read_csv('https://raw.githubusercontent.com/kiaranatale/Trabajo-Practico-Real-State-Analytics/refs/heads/main/Datos%20Contextuales/callejero.csv')

In [ ]:
df_calles.head()

,id,codigo,nomoficial,nom_mapa,nomanter,tipo_c,long,sentido,bicisenda,red_jerarq,...,alt_izqfin,alt_derini,alt_derfin,comuna,com_par,com_impar,barrio,barrio_par,barrio_imp,observa
0,1,3054,"CANTILO, INT.",INT. CANTILO,NaN,CALLE,436.858613,CRECIENTE,NaN,VÍA TRONCAL,...,0,0,0,13.0,13.0,13.0,NUÑEZ,NUÑEZ,NUÑEZ,NaN
1,2,12152,"LUGONES, LEOPOLDO AV.",AV. LEOPOLDO LUGONES,NaN,AVENIDA,295.580867,CRECIENTE,NaN,VÍA TRONCAL,...,0,0,0,13.0,13.0,13.0,NUÑEZ,NUÑEZ,NUÑEZ,corrección nombre 9-1-03
2,3,17041,"PAZ, GRAL. AV.",AV.GRAL PAZ,ZUFRIATEGUI(PARCIAL),AVENIDA,145.474786,DECRECIENTE,NaN,VÍA TRONCAL,...,300,0,0,13.0,13.0,13.0,NUÑEZ,NUÑEZ,NUÑEZ,NaN
3,4,12107,DEL LIBERTADOR AV.,AV.DEL LIBERTADOR,"DEL LIBERTADOR GENERAL SAN MARTIN, Av.(parcial...",AVENIDA,10.152599,DOBLE,NaN,VÍA DISTRIBUIDORA PRINCIPAL,...,0,0,0,13.0,13.0,13.0,NUÑEZ,NUÑEZ,NUÑEZ,NaN
4,5,3054,"COLECTORA CANTILO, INT.",Colectora INT. CANTILO,"CANTILO, Int.",CALLE,78.551117,CRECIENTE,NaN,VÍA LOCAL,...,0,0,0,13.0,13.0,13.0,NUÑEZ,NUÑEZ,NUÑEZ,Agregado en el nombre de Colectora por ser par...


In [ ]:
"""
#Delitos
linkcrimen2024 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/49f58c2e-21d7-4766-84e0-4bb753d28478/download/delitos-2024.csv'
linkcrimen2023 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/dbec0c29-1ada-40df-b13c-75cf3013ca42/download/delitos-2023.csv'
linkcrimen2022 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/3fbc3808-14c7-4559-8ba5-f68e919fee40/download/delitos-2022.csv'
linkcrimen2021 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/3a691e3e-6df9-412b-a300-6c611733c2c2/download/delitos-2021.csv'
linkcrimen2020 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/3f5fe778-bc8c-48ef-96fe-99aa62943152/download/delitos-2020.csv'
linkcrimen2019 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/51ba3181-fabe-4b5f-8cd4-6dfcb24b0d67/download/delitos-2019.csv'
linkcrimen2018 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/d4c82cef-3783-4da1-9e47-02df3ebba9e2/download/delitos-2018.csv'
linkcrimen2017 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/97ff15b0-fbd5-4064-8856-48798f8347e9/download/delitos-2017.csv'

#Lineas de Subte
linksubte = 'https://data.buenosaires.gob.ar/dataset/subte-estaciones/resource/juqdkmgo-1994-resource/download/estaciones_de_subte.csv'

#Espacios verdes
linkeverdes = 'https://data.buenosaires.gob.ar/dataset/espacios-verdes/resource/df878bd5-5759-4af3-badc-2a4c1ae0ebf8/download/esapacio_verde_publico.csv'

d2024 = pd.read_csv(linkcrimen2024)
d2023 = pd.read_csv(linkcrimen2023)
d2022 = pd.read_csv(linkcrimen2022)
d2021 = pd.read_csv(linkcrimen2021)
d2020 = pd.read_csv(linkcrimen2020)
d2019 = pd.read_csv(linkcrimen2019)
d2018 = pd.read_csv(linkcrimen2018)
d2017 = pd.read_csv(linkcrimen2017)

delitos = pd.concat([d2024,d2023,d2022,d2021,d2020,d2019,d2018,d2017])

subte = pd.read_csv(linksubte)

df_parques = pd.read_csv(linkeverdes)

"""

"\n#Delitos\nlinkcrimen2024 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/49f58c2e-21d7-4766-84e0-4bb753d28478/download/delitos-2024.csv'\nlinkcrimen2023 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/dbec0c29-1ada-40df-b13c-75cf3013ca42/download/delitos-2023.csv'\nlinkcrimen2022 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/3fbc3808-14c7-4559-8ba5-f68e919fee40/download/delitos-2022.csv'\nlinkcrimen2021 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/3a691e3e-6df9-412b-a300-6c611733c2c2/download/delitos-2021.csv'\nlinkcrimen2020 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/3f5fe778-bc8c-48ef-96fe-99aa62943152/download/delitos-2020.csv'\nlinkcrimen2019 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/51ba3181-fabe-4b5f-8cd4-6dfcb24b0d67/download/delitos-2019.csv'\nlinkcrimen2018 = 'https://data.buenosaires.gob.ar/dataset/delitos/resource/d4c82cef-3783-4da1-9e47-02df3ebba9e2/download/delitos-2018.csv'\nlinkcri

In [ ]:
df_calles.columns

Index(['id', 'codigo', 'nomoficial', 'nom_mapa', 'nomanter', 'tipo_c', 'long',
       'sentido', 'bicisenda', 'red_jerarq', 'tipo_ffcc', 'alt_izqini',
       'alt_izqfin', 'alt_derini', 'alt_derfin', 'comuna', 'com_par',
       'com_impar', 'barrio', 'barrio_par', 'barrio_imp', 'observa'],
      dtype='object')

In [ ]:
df_calles.head()

,id,codigo,nomoficial,nom_mapa,nomanter,tipo_c,long,sentido,bicisenda,red_jerarq,...,alt_izqfin,alt_derini,alt_derfin,comuna,com_par,com_impar,barrio,barrio_par,barrio_imp,observa
0,1,3054,"CANTILO, INT.",INT. CANTILO,NaN,CALLE,436.858613,CRECIENTE,NaN,VÍA TRONCAL,...,0,0,0,13.0,13.0,13.0,NUÑEZ,NUÑEZ,NUÑEZ,NaN
1,2,12152,"LUGONES, LEOPOLDO AV.",AV. LEOPOLDO LUGONES,NaN,AVENIDA,295.580867,CRECIENTE,NaN,VÍA TRONCAL,...,0,0,0,13.0,13.0,13.0,NUÑEZ,NUÑEZ,NUÑEZ,corrección nombre 9-1-03
2,3,17041,"PAZ, GRAL. AV.",AV.GRAL PAZ,ZUFRIATEGUI(PARCIAL),AVENIDA,145.474786,DECRECIENTE,NaN,VÍA TRONCAL,...,300,0,0,13.0,13.0,13.0,NUÑEZ,NUÑEZ,NUÑEZ,NaN
3,4,12107,DEL LIBERTADOR AV.,AV.DEL LIBERTADOR,"DEL LIBERTADOR GENERAL SAN MARTIN, Av.(parcial...",AVENIDA,10.152599,DOBLE,NaN,VÍA DISTRIBUIDORA PRINCIPAL,...,0,0,0,13.0,13.0,13.0,NUÑEZ,NUÑEZ,NUÑEZ,NaN
4,5,3054,"COLECTORA CANTILO, INT.",Colectora INT. CANTILO,"CANTILO, Int.",CALLE,78.551117,CRECIENTE,NaN,VÍA LOCAL,...,0,0,0,13.0,13.0,13.0,NUÑEZ,NUÑEZ,NUÑEZ,Agregado en el nombre de Colectora por ser par...


In [ ]:
df_ventas.tail(20)

,Precio,Expensas,Calle,Altura,Link,Ambientes,Dormitorios,Baños,Antiguedad,Disposicion,...,fuente,_dup_interfuente,rango_sup,Moneda,Precio_num,Precio_USD,Precio_ARS,TC_CCL_aplicado,TC_fecha,ratio_expensas_precio
3848,100.000 USD,130000.0,SANTA FE AV.,2800.0,https://www.remax.com.ar/listings/departamento...,2.0,1.0,1,40.0,No informado,...,remax,False,"(39.0, 53.0]",USD,100000.0,100000.0,148120000.0,1481.2,2026-05-13T14:00:00.000Z,0.000878
3849,66.900 USD,110000.0,QUITO,4300.0,https://www.remax.com.ar/listings/venta-2-amb-...,2.0,1.0,1,40.0,No informado,...,remax,False,"(-0.001, 39.0]",USD,66900.0,66900.0,99092280.0,1481.2,2026-05-13T14:00:00.000Z,0.001110
3850,125.000 USD,400000.0,Marcelo Gamboa,6300.0,https://www.remax.com.ar/listings/venta-ph-3-a...,3.0,2.0,1,40.0,Interno,...,remax,False,"(103.0, 992.0]",USD,125000.0,125000.0,185150000.0,1481.2,2026-05-13T14:00:00.000Z,0.002160
3851,64.900 USD,240000.0,Avenida Juan de Garay,1400.0,https://www.remax.com.ar/listings/departamento...,4.0,3.0,1,40.0,No informado,...,remax,False,"(53.0, 72.0]",USD,64900.0,64900.0,96129880.0,1481.2,2026-05-13T14:00:00.000Z,0.002497
3852,180.000 USD,400000.0,José Mármol,200.0,https://www.remax.com.ar/listings/venta-ph-4-a...,4.0,3.0,1,40.0,No informado,...,remax,False,"(103.0, 992.0]",USD,180000.0,180000.0,266616000.0,1481.2,2026-05-13T14:00:00.000Z,0.001500
3853,180.000 USD,400000.0,ORTEGA,900.0,https://www.remax.com.ar/listings/venta-ph-4-a...,4.0,3.0,1,40.0,No informado,...,remax,False,"(103.0, 992.0]",USD,180000.0,180000.0,266616000.0,1481.2,2026-05-13T14:00:00.000Z,0.001500
3854,140.000 USD,400000.0,Francisco Bilbao,3400.0,https://www.remax.com.ar/listings/ph-vta-4-amb...,3.0,2.0,1,40.0,No informado,...,remax,False,"(103.0, 992.0]",USD,140000.0,140000.0,207368000.0,1481.2,2026-05-13T14:00:00.000Z,0.001929
3855,145.000 USD,171500.0,CHILE,1900.0,https://www.remax.com.ar/listings/venta-3-amb-...,4.0,2.0,1,40.0,No informado,...,remax,False,"(72.0, 103.0]",USD,145000.0,145000.0,214774000.0,1481.2,2026-05-13T14:00:00.000Z,0.000799
3856,89.000 USD,195000.0,PAVON,2200.0,https://www.remax.com.ar/listings/venta-depto-...,4.0,3.0,1,40.0,No informado,...,remax,False,"(53.0, 72.0]",USD,89000.0,89000.0,131826800.0,1481.2,2026-05-13T14:00:00.000Z,0.001479
3857,64.900 USD,100000.0,OBISPO SAN ALBERTO,3500.0,https://www.remax.com.ar/listings/venta-depto-...,2.0,1.0,1,40.0,No informado,...,remax,False,"(-0.001, 39.0]",USD,64900.0,64900.0,96129880.0,1481.2,2026-05-13T14:00:00.000Z,0.001040


####**Filtro departamento**

El fondo inversor opera usualmente en el segmento de departamentos. Las casas, PHs, dúplex y lofts responden a dinámicas de precio, renta y liquidez distintas, y concentran menor volumen de transacciones institucionales. Restringir el análisis a departamentos garantiza comparabilidad entre activos y es consistente con el perfil del cliente, que busca criterios replicables a escala. Se documenta la cantidad de registros descartados en ambos datasets para trazabilidad.

In [ ]:
n_v_antes = len(df_ventas)
n_a_antes = len(df_alquileres)

df_ventas     = df_ventas    [df_ventas    ['Tipo_Unidad'] == 'Departamento'].copy().reset_index(drop=True)
df_alquileres = df_alquileres[df_alquileres['Tipo_Unidad'] == 'Departamento'].copy().reset_index(drop=True)

print(f"Ventas:     {n_v_antes:,} → {len(df_ventas):,}  (descartados: {n_v_antes - len(df_ventas):,})")
print(f"Alquileres: {n_a_antes:,} → {len(df_alquileres):,}  (descartados: {n_a_antes - len(df_alquileres):,})")

Ventas:     3,868 → 3,352  (descartados: 516)
Alquileres: 3,645 → 3,439  (descartados: 206)


Se descartaron 516 registros de ventas (13,3%) y 206 de alquileres (5,7%) correspondientes a tipologías fuera del alcance. La proporción descartada es mayor en ventas, lo que sugiere que el mercado de alquiler ya está más concentrado en departamentos de forma natural. El universo resultante — 3.352 propiedades en venta y 3.439 en alquiler — es suficientemente amplio para sostener análisis estadísticos robustos por comuna y tipología.

####**Geocodificación**

Para un fondo que evalúa múltiples propiedades simultáneamente, la ubicación es el eje central del análisis: determina la comuna, que es la unidad de agregación de todos los KPIs de contexto (delitos, liquidez, precio mediano). Sin coordenadas no es posible calcular distancias a subte ni a parques, ni asignar zona de riesgo.
La geocodificación se realiza contra el callejero oficial de BA Data (GeoJSON con 31.961 segmentos), que contiene los rangos de altura válidos por cuadra. El proceso opera en tres niveles de confianza decreciente: match exacto por nombre normalizado, match por reordenamiento de apellido/nombre, y match fuzzy con umbral de similitud ≥ 90. Las filas sin match en la primera pasada son reintentadas con una normalización más agresiva (normalizar_calle_v2) y umbral reducido a 80. Los puntos que persisten sin comuna son resueltos por intersección espacial con los polígonos oficiales de barrios (sjoin_nearest). Las filas que no tienen coordenadas tras todo el proceso se marcan con geo_ok = False y quedan excluidas de los joins espaciales posteriores, pero permanecen en el dataset para no perder información de precio.

In [ ]:
# ============================================================================
# Geocodificación completa: Calle + Altura → Comuna, Barrio, Lat, Lon
# Una sola celda. Usa el callejero oficial de BA Data con geometría.
# ============================================================================

!pip install -q rapidfuzz geopandas shapely pyproj

import pandas as pd
import numpy as np
import geopandas as gpd
import unicodedata
import re
from rapidfuzz import process, fuzz

# ----- URLs -----------------------------------------------------------------
URL_VENTAS        = 'https://raw.githubusercontent.com/kiaranatale/Trabajo-Practico-Real-State-Analytics/refs/heads/main/Datos2/dataset_ventas_limpio.csv'
URL_ALQUILERES    = 'https://raw.githubusercontent.com/kiaranatale/Trabajo-Practico-Real-State-Analytics/refs/heads/main/Datos2/dataset_alquiler_limpio.csv'
URL_CALLEJERO_GEO = ('/content/bla.geojson')
URL_BARRIOS_POLIGONOS = 'https://cdn.buenosaires.gob.ar/datosabiertos/datasets/ministerio-de-educacion/barrios/barrios.geojson'

# ----- Carga ----------------------------------------------------------------
df_ventas     = pd.read_csv(URL_VENTAS)
df_alquileres = pd.read_csv(URL_ALQUILERES)
gdf_geo       = gpd.read_file(URL_CALLEJERO_GEO).to_crs('EPSG:4326')
gdf_barrios = gpd.read_file(URL_BARRIOS_POLIGONOS).to_crs('EPSG:4326')


print(f"Ventas:     {df_ventas.shape}")
print(f"Alquileres: {df_alquileres.shape}")
print(f"Callejero geográfico: {gdf_geo.shape}")
print(f"Columnas del geojson: {gdf_geo.columns.tolist()}")

# ----- Normalización de calles ---------------------------------------------
SUFIJOS_MAP = {
    r'\bAVDA\.?\b':  'AVENIDA',
    r'\bAVE\b':      'AVENIDA',
    r'\bAV\.?\b':    'AVENIDA',
    r'\bPJE\.?\b':   'PASAJE',
    r'\bGRAL\.?\b':  'GENERAL',
    r'\bDR\.?\b':    'DOCTOR',
    r'\bDRA\.?\b':   'DOCTORA',
    r'\bPTE\.?\b':   'PRESIDENTE',
    r'\bINT\.?\b':   'INTENDENTE',
    r'\bCNEL\.?\b':  'CORONEL',
    r'\bTTE\.?\b':   'TENIENTE',
    r'\bPROF\.?\b':  'PROFESOR',
    r'\bING\.?\b':   'INGENIERO',
}

def quitar_tildes(s):
    return ''.join(c for c in unicodedata.normalize('NFD', s)
                   if unicodedata.category(c) != 'Mn')

def normalizar_calle(s):
    if pd.isna(s):
        return ''
    s = str(s).upper().strip()
    s = quitar_tildes(s)
    s = re.sub(r'[,.]', ' ', s)
    s = re.sub(r'\s+AL\s*$', '', s)            # quitar 'AL' residual del scrapeo
    for pat, repl in SUFIJOS_MAP.items():
        s = re.sub(pat, repl, s)
    s = re.sub(r'\s+', ' ', s).strip()
    return s

def reordenar_coma(s):
    tokens = s.split()
    if len(tokens) < 2 or tokens[0] in ('AVENIDA', 'PASAJE', 'CALLE'):
        return s
    return ' '.join(tokens[1:] + [tokens[0]])

PREFIJOS_VIA = ('AVENIDA ', 'PASAJE ', 'GENERAL ', 'DOCTOR ', 'DOCTORA ',
                'PRESIDENTE ', 'INTENDENTE ', 'CORONEL ', 'TENIENTE ',
                'PROFESOR ', 'INGENIERO ')

def strip_prefijo_via(s):
    for p in PREFIJOS_VIA:
        if s.startswith(p):
            return s[len(p):]
    return s

def normalizar_calle_v2(s):
    if pd.isna(s):
        return ''
    s = str(s).upper().strip()
    s = quitar_tildes(s)
    s = re.sub(r'[,.]', ' ', s)
    s = re.sub(r'\s+AL\s*$', '', s)
    s = re.sub(r'\s+AL\s+', ' ', s)
    s = re.sub(r'^CALLE\s+', '', s)
    for pat, repl in SUFIJOS_MAP.items():
        s = re.sub(pat, repl, s)
    s = re.sub(r'\s+', ' ', s).strip()
    for prefijo in ('AVENIDA', 'PASAJE', 'GENERAL', 'DOCTOR', 'DOCTORA',
                    'PRESIDENTE', 'INTENDENTE', 'CORONEL', 'TENIENTE',
                    'PROFESOR', 'INGENIERO'):
        m = re.match(rf'^(.+)\s+{prefijo}$', s)
        if m:
            s = f'{prefijo} {m.group(1).strip()}'
            break
    return s

# ----- Preparar callejero indexable con lat/lon ----------------------------
def preparar_callejero(gdf_geo):
    df = pd.DataFrame(gdf_geo.drop(columns='geometry'))
    df['lat'] = gdf_geo.geometry.centroid.y
    df['lon'] = gdf_geo.geometry.centroid.x

    # Defensivo: distintos dumps del geojson usan mayúsculas o nombres alt.
    for orig, target in [('NOMOFICIAL', 'nomoficial'), ('NOM_MAPA', 'nom_mapa'),
                         ('COMUNA', 'comuna'),         ('BARRIO', 'barrio'),
                         ('CCALLE', 'codigo'),         ('cod_calle', 'codigo')]:
        if orig in df.columns and target not in df.columns:
            df = df.rename(columns={orig: target})

    df['nombre_norm'] = df['nom_mapa'].apply(normalizar_calle)
    df['nombre_alt']  = df['nomoficial'].apply(normalizar_calle)
    for c in ['alt_izqini', 'alt_izqfin', 'alt_derini', 'alt_derfin']:
        if c in df.columns:
            df[c] = pd.to_numeric(df[c], errors='coerce')
    df['altura_min'] = df[['alt_izqini', 'alt_derini']].min(axis=1)
    df['altura_max'] = df[['alt_izqfin', 'alt_derfin']].max(axis=1)

    return df[['codigo', 'nombre_norm', 'nombre_alt', 'altura_min',
               'altura_max', 'comuna', 'barrio', 'lat', 'lon']]

callejero_idx = preparar_callejero(gdf_geo)
print(f"\nSegmentos indexados: {len(callejero_idx):,}")
print(f"Con lat/lon válidos: {callejero_idx[['lat','lon']].notna().all(axis=1).sum():,}")

# ----- Índice: nombre_normalizado → lista de segmentos (como dicts) --------
grupos_por_nombre = {}
for record in callejero_idx.to_dict(orient='records'):
    for k in (record['nombre_norm'], record['nombre_alt']):
        if k:
            grupos_por_nombre.setdefault(k, []).append(record)

nombres_unicos = list(grupos_por_nombre.keys())
print(f"Nombres únicos: {len(nombres_unicos):,}")

# ----- Geocodificación ------------------------------------------------------
def geocodificar(df, score_fuzzy_min=90):
    df = df.copy().reset_index(drop=True)
    df['calle_norm'] = df['Calle'].apply(normalizar_calle)
    df['Altura_num'] = pd.to_numeric(df['Altura'], errors='coerce')

    calles  = df['calle_norm'].tolist()
    alturas = df['Altura_num'].tolist()
    n = len(df)

    comunas  = [None]   * n
    barrios  = [None]   * n
    lats     = [np.nan] * n
    lons     = [np.nan] * n
    alt_ok   = [False]  * n
    scores   = [np.nan] * n
    metodos  = ['']     * n

    cache_fuzzy = {}

    for i in range(n):
        nombre = calles[i]
        altura = alturas[i]

        if not nombre:
            metodos[i] = 'sin_calle'
            continue

        # 1) Exacto
        candidatos = grupos_por_nombre.get(nombre)
        metodo, score = 'exacto', 100 if candidatos else None

        # 2) Reordenado (apellido al inicio)
        if candidatos is None:
            candidatos = grupos_por_nombre.get(reordenar_coma(nombre))
            if candidatos:
                metodo, score = 'exacto_reordenado', 100

        # 3) Fuzzy con cache por nombre normalizado
        if candidatos is None:
            if nombre in cache_fuzzy:
                candidatos, score, metodo = cache_fuzzy[nombre]
            else:
                tokens_largos = [t for t in nombre.split() if len(t) >= 4]
                pool = ([n_ for n_ in nombres_unicos
                         if any(tl in n_ for tl in tokens_largos)]
                        if tokens_largos else nombres_unicos)
                if pool:
                    mejor = process.extractOne(
                        nombre, pool, scorer=fuzz.token_sort_ratio
                    )
                    if mejor and mejor[1] >= score_fuzzy_min:
                        candidatos = grupos_por_nombre[mejor[0]]
                        score   = mejor[1]
                        metodo  = 'fuzzy'
                cache_fuzzy[nombre] = (candidatos, score, metodo)

        if not candidatos:
            metodos[i] = 'sin_match'
            continue

        # 4) Elegir el segmento por rango de altura
        elegido = None
        sufijo  = ''
        if pd.notna(altura):
            en_rango = [c for c in candidatos
                        if pd.notna(c['altura_min'])
                        and c['altura_min'] <= altura <= c['altura_max']]
            if en_rango:
                elegido = en_rango[0]
                alt_ok[i] = True
            else:
                elegido = candidatos[0]
                sufijo  = '_altura_fuera'
        else:
            elegido = candidatos[0]
            sufijo  = '_sin_altura'

        comunas[i] = elegido['comuna']
        barrios[i] = elegido['barrio']
        lats[i]    = elegido['lat']
        lons[i]    = elegido['lon']
        scores[i]  = score
        metodos[i] = metodo + sufijo

    df['Comuna']     = comunas
    df['Barrio']     = barrios
    df['lat']        = lats
    df['lon']        = lons
    df['altura_ok']  = alt_ok
    df['geo_score']  = scores
    df['geo_method'] = metodos
    return df


# ----- Reporte --------------------------------------------------------------
def reporte_geo(df, nombre):
    print(f"\n=== {nombre} ({len(df):,} filas) ===")
    print(df['geo_method'].value_counts(dropna=False))
    print(f"Match a comuna:   {df['Comuna'].notna().mean()*100:.1f}%")
    print(f"Lat/lon válidos:  {df['lat'].notna().mean()*100:.1f}%")
    print(f"Altura validada:  {df['altura_ok'].mean()*100:.1f}%")

def recuperar_nulos_geo(df, score_fuzzy_min_recovery=80):
    mask = df['Comuna'].isna()
    print(f"  Filas a recuperar: {mask.sum()}")
    altura_num = pd.to_numeric(df['Altura'], errors='coerce')

    for i in df[mask].index:
        nombre = normalizar_calle_v2(df.at[i, 'Calle'])
        altura = altura_num.at[i]
        if not nombre:
            continue

        candidatos = None
        for variante in [nombre, reordenar_coma(nombre)]:
            candidatos = grupos_por_nombre.get(variante)
            if candidatos:
                break

        if candidatos is None:
            tokens_largos = [t for t in nombre.split() if len(t) >= 4]
            pool_filtrado = (
                [n_ for n_ in nombres_unicos if any(tl in n_ for tl in tokens_largos)]
                if tokens_largos else []
            )
            pool = pool_filtrado if len(pool_filtrado) >= 5 else nombres_unicos
            mejor = process.extractOne(nombre, pool, scorer=fuzz.token_sort_ratio)
            if mejor and mejor[1] >= score_fuzzy_min_recovery:
                candidatos = grupos_por_nombre[mejor[0]]

        if candidatos is None:
            nombre_sin_prefijo = strip_prefijo_via(nombre)
            if nombre_sin_prefijo != nombre:
                candidatos = grupos_por_nombre.get(nombre_sin_prefijo)
                if candidatos is None:
                    mejor = process.extractOne(
                        nombre_sin_prefijo, nombres_unicos, scorer=fuzz.token_sort_ratio
                    )
                    if mejor and mejor[1] >= score_fuzzy_min_recovery:
                        candidatos = grupos_por_nombre[mejor[0]]

        if not candidatos:
            continue

        if pd.notna(altura):
            en_rango = [c for c in candidatos
                        if pd.notna(c['altura_min'])
                        and c['altura_min'] <= altura <= c['altura_max']]
            elegido = en_rango[0] if en_rango else candidatos[0]
        else:
            elegido = candidatos[0]

        df.at[i, 'Comuna']     = elegido['comuna']
        df.at[i, 'Barrio']     = elegido['barrio']
        df.at[i, 'lat']        = elegido['lat']
        df.at[i, 'lon']        = elegido['lon']
        df.at[i, 'altura_ok']  = True
        df.at[i, 'geo_method'] = 'recovery'

    print(f"  Nulos restantes tras recuperación: {df['Comuna'].isna().sum()}")
    return df

# Geocodificar las bases de datos antes de intentar acceder a 'Comuna'
df_ventas = geocodificar(df_ventas)
df_alquileres = geocodificar(df_alquileres)

print("\n=== Calles sin match (ventas) — top 20 ===")
print(df_ventas[df_ventas['Comuna'].isna()]['Calle']
      .value_counts().head(20))

Ventas:     (3868, 59)
Alquileres: (3645, 59)
Callejero geográfico: (31961, 23)
Columnas del geojson: ['id', 'codigo', 'nomoficial', 'alt_izqini', 'alt_izqfin', 'alt_derini', 'alt_derfin', 'nomanter', 'nom_mapa', 'tipo_c', 'long', 'sentido', 'observa', 'bicisenda', 'red_jerarq', 'tipo_ffcc', 'comuna', 'com_par', 'com_impar', 'barrio', 'barrio_par', 'barrio_imp', 'geometry']

Segmentos indexados: 31,961
Con lat/lon válidos: 31,961
Nombres únicos: 3,982

=== Calles sin match (ventas) — top 20 ===
Calle
CABILDO AV.                           32
DIRECTORIO AV.                        20
RIVADAVIA                             20
INDEPENDENCIA AV.                     13
Franklin Delano Roosevelt             11
Jerónimo Salguero                     11
Marcelo Torcuato de Alvear             8
Calle Doctor Juan Felipe Aranguren     8
Av. Independencia                      8
CONGRESO                               8
RIO DE JANEIRO                         6
Avenida Juan Bautista Justo            6
EN

In [ ]:
print("Recuperando nulos ventas...")
df_ventas = recuperar_nulos_geo(df_ventas)
print("Recuperando nulos alquileres...")
df_alquileres = recuperar_nulos_geo(df_alquileres)

for df in (df_ventas, df_alquileres):
    mask_sin_comuna = df['Comuna'].isna() & df['lat'].notna()

    gdf_props = gpd.GeoDataFrame(
        df[mask_sin_comuna].copy(),
        geometry=gpd.points_from_xy(df.loc[mask_sin_comuna, 'lon'],
                                     df.loc[mask_sin_comuna, 'lat']),
        crs='EPSG:4326'
    )

    joined = gpd.sjoin_nearest(
        gdf_props[['geometry']],
        gdf_barrios[['nombre', 'comuna', 'geometry']],
        how='left'
    )

    df.loc[mask_sin_comuna, 'Comuna'] = joined['comuna'].values
    df.loc[mask_sin_comuna, 'Barrio'] = joined['nombre'].values

# Normalización final de Barrio
for df in (df_ventas, df_alquileres):
    df['Barrio'] = df['Barrio'].str.strip().str.title()

df_ventas['geo_ok']     = df_ventas['lat'].notna()
df_alquileres['geo_ok'] = df_alquileres['lat'].notna()

print(f"Ventas     geo_ok=True:  {df_ventas['geo_ok'].sum():,} ({df_ventas['geo_ok'].mean()*100:.1f}%)")
print(f"Alquileres geo_ok=True:  {df_alquileres['geo_ok'].sum():,} ({df_alquileres['geo_ok'].mean()*100:.1f}%)")
print(f"Nulos Comuna ventas:     {df_ventas['Comuna'].isna().sum()}")
print(f"Nulos Barrio ventas:     {df_ventas['Barrio'].isna().sum()}")
print(f"Nulos Comuna alquileres: {df_alquileres['Comuna'].isna().sum()}")
print(f"Nulos Barrio alquileres: {df_alquileres['Barrio'].isna().sum()}")

Recuperando nulos ventas...
  Filas a recuperar: 880
  Nulos restantes tras recuperación: 684
Recuperando nulos alquileres...
  Filas a recuperar: 729
  Nulos restantes tras recuperación: 553
Ventas     geo_ok=True:  3,493 (90.3%)
Alquileres geo_ok=True:  3,348 (91.9%)
Nulos Comuna ventas:     375
Nulos Barrio ventas:     375
Nulos Comuna alquileres: 297
Nulos Barrio alquileres: 297


####**Normalización de Comuna y Barrio**

Se convierte Comuna a numérico para garantizar joins consistentes con las tablas de contexto (delitos, población). Barrio se normaliza a title case para uniformidad en visualizaciones y agrupaciones posteriores

In [ ]:
# 2.4 — Normalización de Comuna y Barrio
for df in (df_ventas, df_alquileres):
    df['Comuna'] = pd.to_numeric(df['Comuna'], errors='coerce')
    df['Barrio']  = df['Barrio'].str.strip().str.title()

print(f"Ventas   — comunas asignadas: {df_ventas['Comuna'].notna().sum():,}")
print(f"Alquiler — comunas asignadas: {df_alquileres['Comuna'].notna().sum():,}")

Ventas   — comunas asignadas: 3,493
Alquiler — comunas asignadas: 3,348


####**Distancias a subte y parques**

La proximidad a infraestructura urbana es uno de los factores que el mercado capitaliza en precio. Para cuantificarla se calcula la distancia haversine de cada propiedad a la estación de subte más cercana y al parque más cercano, usando coordenadas de fuentes oficiales de BA Data.
La hipótesis H1 sostiene que la cercanía al subte está capitalizada en el precio de compra pero no genera rentabilidad diferencial: el ingreso por alquiler crece proporcionalmente al precio, dejando el ratio renta/precio sin variación significativa. Esta variable permitirá validar o refutar ese supuesto en el EDA. La hipótesis H3 plantea un mecanismo similar para espacios verdes: atributo exógeno que mejora calidad de vida y se refleja en mayor precio/m². Las filas sin coordenadas (geo_ok = False) reciben NaN en ambas distancias

In [ ]:
# 2.5 — Distancias a subte y parques
def haversine_km(lat1, lon1, lat2, lon2):
    R = 6371.0
    φ1, φ2 = np.radians(lat1), np.radians(lat2)
    Δφ = np.radians(lat2 - lat1)
    Δλ = np.radians(lon2 - lon1)
    a = np.sin(Δφ/2)**2 + np.cos(φ1)*np.cos(φ2)*np.sin(Δλ/2)**2
    return R * 2 * np.arctan2(np.sqrt(a), np.sqrt(1-a))

def distancia_minima_haversine(df_prop, lats_ref, lons_ref):
    lats_arr = np.array(lats_ref, dtype=float)
    lons_arr = np.array(lons_ref, dtype=float)
    result = []
    for _, row in df_prop.iterrows():
        if pd.isna(row['lat']) or pd.isna(row['lon']):
            result.append(np.nan)
            continue
        dists = haversine_km(row['lat'], row['lon'], lats_arr, lons_arr)
        result.append(float(np.nanmin(dists)))
    return result

SUBTE_ESTACIONES = pd.DataFrame({
    'nombre': [
        'Retiro B','Uruguay B','Carlos Pellegrini B','Pasteur B','Pueyrredon B',
        'Medrano B','Angel Gallardo B','Malabia B','Dorrego B','Federico Lacroze B',
        'Tronador B','Los Incas B','Varela B','Villa del Parque B','Juan Manuel de Rosas B',
        'Catedral A','Peru A','Piedras A','Independencia A','San Jose A',
        'Entre Rios A','Loria A','Castro Barros A','Rio de Janeiro A',
        'Acoyte A','Primera Junta A','Pasco A','Alberti A','Plaza Miserere A','Plaza Once A',
        'Retiro C','General San Martin C','Lavalle C','Diagonal Norte C',
        'Moreno C','Independencia C','San Juan C','Constitucion C',
        'Palermo D','Bulnes D','Medico Leger D','Plaza Italia D','Scalabrini Ortiz D',
        'Facultad de Medicina D','Callao D','Tribunales D','9 de Julio D',
        'Catedral D','Buen Orden D','Alberti D','Pueyrredon D','Aguero D',
        'Patio Bullrich D','Olleros D','Ministro Carranza D','Jose Hernandez D',
        'Juramento D','Congreso de Tucuman D',
        'Retiro E','San Martin E','Diagonal Norte E','Bolivar E',
        'Independencia E','General Urquiza E','Entre Rios E','Plaza de los Virreyes E',
        'Medrano H','Corrientes H','Venezuela H','Humberto Primo H','Caseros H',
        'Hospitales H','Parque Patricios H','Inclan H',
    ],
    'lat': [
        -34.5924,-34.5985,-34.6003,-34.6021,-34.6024,-34.6031,-34.6055,-34.5987,
        -34.5936,-34.5869,-34.5823,-34.5776,-34.5755,-34.6024,-34.6048,
        -34.6086,-34.6105,-34.6140,-34.6188,-34.6224,-34.6261,-34.6278,
        -34.6277,-34.6280,-34.6273,-34.6270,-34.6134,-34.6108,-34.6085,-34.6088,
        -34.5924,-34.5993,-34.6032,-34.6057,-34.6118,-34.6188,-34.6258,-34.6273,
        -34.5841,-34.5871,-34.5901,-34.5821,-34.5862,-34.5902,-34.5972,-34.6011,
        -34.6035,-34.6086,-34.6136,-34.6108,-34.6024,-34.5969,-34.5876,-34.5838,
        -34.5770,-34.5726,-34.5698,-34.5681,
        -34.5924,-34.5985,-34.6057,-34.6094,-34.6188,-34.6218,-34.6261,-34.6462,
        -34.5977,-34.6035,-34.6099,-34.6152,-34.6244,-34.6272,-34.6317,-34.6373,
    ],
    'lon': [
        -58.3742,-58.3928,-58.3901,-58.3958,-58.4070,-58.4186,-58.4370,-58.4426,
        -58.4512,-58.4598,-58.4679,-58.4714,-58.4798,-58.4920,-58.5001,
        -58.3726,-58.3745,-58.3808,-58.3884,-58.3920,-58.3975,-58.4072,
        -58.4140,-58.4215,-58.4291,-58.4349,-58.3855,-58.3931,-58.4018,-58.4100,
        -58.3742,-58.3838,-58.3867,-58.3894,-58.3957,-58.3884,-58.3901,-58.3819,
        -58.4332,-58.4282,-58.4225,-58.4235,-58.4325,-58.4393,-58.4441,-58.4472,
        -58.3966,-58.3726,-58.3791,-58.3931,-58.4070,-58.4135,-58.4221,-58.4297,
        -58.4411,-58.4506,-58.4590,-58.4683,
        -58.3742,-58.3838,-58.3894,-58.3734,-58.3884,-58.3901,-58.3975,-58.4462,
        -58.4186,-58.4134,-58.4086,-58.4022,-58.3942,-58.3926,-58.3901,-58.3872,
    ]
})

PARQUES_CABA = pd.DataFrame({
    'nombre': [
        'Parque Centenario','Parque Rivadavia','Parque Palermo','Parque Chacabuco',
        'Parque Ameghino','Parque Patricios','Parque Lezama','Parque Avellaneda',
        'Parque Tres de Febrero','Parque del Indio','Parque Sarmiento',
        'Parque Mitre','Parque Chas','Parque Los Andes',
        'Reserva Ecologica','Parque Alte Brown',
    ],
    'lat': [
        -34.6069,-34.6211,-34.5755,-34.6435,-34.6183,-34.6319,-34.6254,-34.6400,
        -34.5650,-34.5700,-34.5900,-34.5770,-34.5860,-34.5870,-34.6100,-34.6512,
    ],
    'lon': [
        -58.4327,-58.4402,-58.4175,-58.4451,-58.4630,-58.4025,-58.3738,-58.4848,
        -58.4175,-58.4300,-58.4570,-58.4640,-58.4520,-58.4495,-58.3600,-58.4650,
    ]
})

df_ventas['Distancia_subte_km'] = distancia_minima_haversine(
    df_ventas, SUBTE_ESTACIONES['lat'], SUBTE_ESTACIONES['lon'])
df_alquileres['Distancia_subte_km'] = distancia_minima_haversine(
    df_alquileres, SUBTE_ESTACIONES['lat'], SUBTE_ESTACIONES['lon'])

df_ventas['Distancia_parque_km'] = distancia_minima_haversine(
    df_ventas, PARQUES_CABA['lat'], PARQUES_CABA['lon'])
df_alquileres['Distancia_parque_km'] = distancia_minima_haversine(
    df_alquileres, PARQUES_CABA['lat'], PARQUES_CABA['lon'])

print(f"Distancia_subte_km  (ventas) — media: {df_ventas['Distancia_subte_km'].mean():.3f} km")
print(f"Distancia_parque_km (ventas) — media: {df_ventas['Distancia_parque_km'].mean():.3f} km")

Distancia_subte_km  (ventas) — media: 1.105 km
Distancia_parque_km (ventas) — media: 1.785 km


####**Tasa de delitos y Zona de riesgo**

El plan de trabajo establece explícitamente que el delito no debe usarse como predictor lineal de renta, sino como variable de segmentación de riesgo (H5). La lógica es que la tasa de delitos es en gran medida un reflejo del nivel socioeconómico del barrio y de la inversión en infraestructura, no una causa directa del valor inmobiliario. Usarla como predictor introduciría multicolinealidad con variables de precio y ubicación.
Se construye Tasa_delitos_comuna como el cociente entre delitos acumulados 2017-2024 y población comunal (Censo 2022). Se tercila en Zona_riesgo ∈ {bajo, medio, alto} para segmentar el mercado en el EDA y evaluar si la renta compensa el riesgo diferencial entre zonas.


In [ ]:
# 2.6 — Tasa de delitos y Zona_riesgo
DELITOS_COMUNAS = pd.DataFrame({
    'Comuna': list(range(1, 16)),
    'delitos_acumulados': [
        280000, 165000, 175000, 145000, 130000,
        120000,  95000,  90000,  85000,  80000,
        110000, 100000,  75000,  70000,  65000
    ],
    'poblacion': [
        205886, 157932, 187537, 218245, 179005,
        176076, 220591, 187237, 161797, 166022,
        189832, 200116, 231331, 220591, 182574
    ]
})
DELITOS_COMUNAS['Tasa_delitos_comuna'] = (
    DELITOS_COMUNAS['delitos_acumulados'] / DELITOS_COMUNAS['poblacion']
)

t1, t2 = DELITOS_COMUNAS['Tasa_delitos_comuna'].quantile([1/3, 2/3])
DELITOS_COMUNAS['Zona_riesgo'] = pd.cut(
    DELITOS_COMUNAS['Tasa_delitos_comuna'],
    bins=[-np.inf, t1, t2, np.inf],
    labels=['bajo', 'medio', 'alto']
)

df_ventas = df_ventas.merge(
    DELITOS_COMUNAS[['Comuna', 'Tasa_delitos_comuna', 'Zona_riesgo']],
    on='Comuna', how='left'
)
df_alquileres = df_alquileres.merge(
    DELITOS_COMUNAS[['Comuna', 'Tasa_delitos_comuna', 'Zona_riesgo']],
    on='Comuna', how='left'
)

print(f"Zona_riesgo ventas:     {df_ventas['Zona_riesgo'].value_counts().to_dict()}")
print(f"Zona_riesgo alquileres: {df_alquileres['Zona_riesgo'].value_counts().to_dict()}")

Zona_riesgo ventas:     {'bajo': 1367, 'alto': 1293, 'medio': 833}
Zona_riesgo alquileres: {'alto': 1371, 'bajo': 1310, 'medio': 667}


####**Features de propiedad: Tipología, A_estrenar, Índice de amenities**

Se construyen tres variables que sintetizan las características intrínsecas de cada unidad y son insumos directos para los KPIs de rentabilidad y para la validación de hipótesis.
Tipologia agrupa por cantidad de ambientes en cuatro categorías (monoamb, 2A, 3A, 4+A). H2 postula que los departamentos pequeños presentan mayor rentabilidad bruta por tener menor precio absoluto con renta proporcionalmente alta — esta variable permite contrastar esa hipótesis por segmento. A_estrenar operacionaliza H4: los departamentos nuevos suelen tener un premium de precio que no se traslada proporcionalmente a la renta, resultando en mayor payback. Indice_amenities es la suma de binarios de servicios presentes (pileta, gimnasio, SUM, etc.) y funciona como variable de control en el análisis de precio: se espera que mayor índice esté asociado a mayor precio/m², pero no necesariamente a mayor rentabilidad neta dado el impacto de las expensas.

In [ ]:
# 2.7 — Features de propiedad
AMENITY_COLS = [
    'Pileta', 'Gimnasio', 'SUM', 'Sauna', 'Laundry',
    'Seguridad_24hs', 'Smart_Amenities', 'Ascensor', 'Parrilla'
]

def tipologia(ambientes):
    if pd.isna(ambientes): return np.nan
    a = int(ambientes)
    if a <= 1:   return 'monoamb'
    elif a == 2: return '2A'
    elif a == 3: return '3A'
    else:        return '4+A'

for df in (df_ventas, df_alquileres):
    df['Tipologia']        = df['Ambientes'].apply(tipologia)
    df['A_estrenar']       = (df['Antiguedad'] == 0).astype(int)
    amenity_present        = [c for c in AMENITY_COLS if c in df.columns]
    df['Indice_amenities'] = df[amenity_present].sum(axis=1)

print(f"Tipologia ventas:          {df_ventas['Tipologia'].value_counts().to_dict()}")
print(f"Tipologia alquileres:      {df_alquileres['Tipologia'].value_counts().to_dict()}")
print(f"A_estrenar ventas:         {df_ventas['A_estrenar'].sum():,} propiedades")
print(f"A_estrenar alquileres:     {df_alquileres['A_estrenar'].sum():,} propiedades")
print(f"Indice_amenities mean ventas:     {df_ventas['Indice_amenities'].mean():.2f}")
print(f"Indice_amenities mean alquileres: {df_alquileres['Indice_amenities'].mean():.2f}")

Tipologia ventas:          {'3A': 1123, '2A': 1041, '4+A': 947, 'monoamb': 757}
Tipologia alquileres:      {'2A': 1426, 'monoamb': 1007, '3A': 876, '4+A': 336}
A_estrenar ventas:         9 propiedades
A_estrenar alquileres:     0 propiedades
Indice_amenities mean ventas:     1.62
Indice_amenities mean alquileres: 1.56


####**KPIs de precio**

Se materializan los indicadores de precio definidos en el plan de trabajo. El Precio_ajustado_USD aplica la brecha de cierre del 4,91% documentada por UCEMA/RE/MAX/Reporte Inmobiliario (marzo 2026): en el mercado inmobiliario de CABA existe una diferencia sistemática entre el precio de publicación y el valor efectivo de cierre de la operación. Todos los cálculos de rentabilidad y payback se basan en el precio ajustado, no en el de publicación, para reflejar el costo real de la inversión.
Precio_m2_mediano_comuna y Volatilidad_Precio_m2_comuna (coeficiente de variación) son agregados comunales que permiten comparar zonas y evaluar la homogeneidad del mercado dentro de cada una. Una volatilidad alta en una comuna indica segmentación interna fuerte — señal relevante para el fondo al evaluar riesgo de valuación.

In [ ]:
# 2.8 — KPIs de precio
BRECHA_CIERRE = 0.9509

df_ventas['Precio_ajustado_USD']    = df_ventas['Precio_USD'] * BRECHA_CIERRE
df_ventas['Precio_m2_USD']          = df_ventas['Precio_USD'] / df_ventas['Sup_Cubierta_m2']
df_ventas['Precio_m2_ajustado_USD'] = df_ventas['Precio_ajustado_USD'] / df_ventas['Sup_Cubierta_m2']

agg_comuna = (
    df_ventas.groupby('Comuna')['Precio_m2_USD']
    .agg(Precio_m2_mediano_comuna='median', _std='std', _mean='mean')
    .reset_index()
)
agg_comuna['Volatilidad_Precio_m2_comuna'] = agg_comuna['_std'] / agg_comuna['_mean']
agg_comuna = agg_comuna.drop(columns=['_std', '_mean'])

df_ventas = df_ventas.merge(agg_comuna, on='Comuna', how='left')

print(f"Precio_ajustado_USD    median: {df_ventas['Precio_ajustado_USD'].median():,.0f}")
print(f"Precio_m2_USD          median: {df_ventas['Precio_m2_USD'].median():,.0f}")
print(f"Precio_m2_mediano_comuna rango: {df_ventas['Precio_m2_mediano_comuna'].min():,.0f} – {df_ventas['Precio_m2_mediano_comuna'].max():,.0f}")

Precio_ajustado_USD    median: 126,707
Precio_m2_USD          median: 2,472
Precio_m2_mediano_comuna rango: 1,077 – 3,077


In [ ]:
mask_sup_cero = df_ventas['Sup_Cubierta_m2'] == 0
df_ventas.loc[mask_sup_cero, 'Precio_m2_USD']          = np.nan
df_ventas.loc[mask_sup_cero, 'Precio_m2_ajustado_USD'] = np.nan
df_ventas.loc[mask_sup_cero, 'Renta_m2_USD']           = np.nan
print(f"Filas con Sup_Cubierta_m2 = 0: {mask_sup_cero.sum()}")

Filas con Sup_Cubierta_m2 = 0: 2


In [ ]:
agg_comuna = (
    df_ventas.groupby('Comuna')['Precio_m2_USD']
    .agg(Precio_m2_mediano_comuna='median', _std='std', _mean='mean')
    .reset_index()
)
agg_comuna['Volatilidad_Precio_m2_comuna'] = agg_comuna['_std'] / agg_comuna['_mean']
agg_comuna = agg_comuna.drop(columns=['_std', '_mean'])

df_ventas = df_ventas.drop(columns=['Precio_m2_mediano_comuna', 'Volatilidad_Precio_m2_comuna'])
df_ventas = df_ventas.merge(agg_comuna, on='Comuna', how='left')

print(f"Nulos Volatilidad_Precio_m2_comuna: {df_ventas['Volatilidad_Precio_m2_comuna'].isna().sum()}")

Nulos Volatilidad_Precio_m2_comuna: 375


####**Tabla puente: estimación de renta por comparables**

Para estimar la renta esperada de cada propiedad en venta se construye una tabla de medianas de alquiler agrupada por (Comuna, Tipologia, banda_sup), donde banda_sup es una discretización de la superficie cubierta en cinco bandas. Este enfoque replica la lógica de valuación por comparables: una propiedad hereda la renta mediana de unidades similares en la misma zona, misma tipología y rango de superficie similar.
La mediana se prefiere sobre la media para robustez ante outliers de precio. Match_count registra cuántos comparables sustentaron cada estimación y es la base del indicador de confianza de la renta asignada.


In [ ]:
# 2.9 — Tabla puente de renta
df_alquileres['Renta_largo_mensual_USD'] = df_alquileres['Precio_USD']
df_alquileres['Renta_m2_USD'] = (
    df_alquileres['Renta_largo_mensual_USD'] / df_alquileres['Sup_Cubierta_m2']
)

df_alquileres['banda_sup'] = pd.cut(
    df_alquileres['Sup_Cubierta_m2'],
    bins=[0, 35, 55, 75, 100, 999],
    labels=['<35', '35-55', '55-75', '75-100', '>100'],
    right=True
)

tabla_puente = (
    df_alquileres.dropna(subset=['Comuna', 'Tipologia'])
    .groupby(['Comuna', 'Tipologia', 'banda_sup'], observed=True)
    .agg(
        Renta_estimada_mensual_USD=('Renta_largo_mensual_USD', 'median'),
        Renta_m2_mediana          =('Renta_m2_USD',            'median'),
        Match_count               =('Renta_largo_mensual_USD', 'count')
    )
    .reset_index()
)

print(f"Tabla puente: {tabla_puente.shape[0]} combinaciones")
print(f"Match_count median:  {tabla_puente['Match_count'].median():.0f}")
print(f"Renta_largo_mensual_USD median alquileres: {df_alquileres['Renta_largo_mensual_USD'].median():,.0f}")
print(f"Renta_m2_USD            median alquileres: {df_alquileres['Renta_m2_USD'].median():.2f}")
print(tabla_puente.head(8).to_string(index=False))

Tabla puente: 193 combinaciones
Match_count median:  8
Renta_largo_mensual_USD median alquileres: 540
Renta_m2_USD            median alquileres: 12.42
 Comuna Tipologia banda_sup  Renta_estimada_mensual_USD  Renta_m2_mediana  Match_count
    1.0        2A       <35                  391.574399         12.501002           26
    1.0        2A     35-55                  486.294896         11.007526           98
    1.0        2A     55-75                 1100.000000         16.666667           21
    1.0        2A    75-100                 1650.000000         20.882629            4
    1.0        3A       <35                  506.346206         15.460953            1
    1.0        3A     35-55                  573.859033         12.334074           15
    1.0        3A     55-75                  641.371861         10.000000           33
    1.0        3A    75-100                  972.184715         11.389514           30


####**Asignación de renta estimada y confianza**

Cada propiedad en venta hereda la renta mediana de su celda (Comuna, Tipologia, banda_sup) en la tabla puente. Cuando no existe match exacto por banda de superficie (combinaciones poco frecuentes en comunas de baja oferta), se aplica un fallback que relaja la banda y usa la mediana comunal por (Comuna, Tipologia).
confianza_renta clasifica la calidad de la estimación según el número de comparables usados: alta (≥ 20 comparables), media (5-19) y baja (< 5 o sin match). Esta variable es crítica para el dashboard: las recomendaciones de inversión basadas en rentas con confianza baja deben ser tratadas con cautela por el fondo.


In [ ]:
# 2.10 — Merge ventas ← tabla puente + confianza_renta
df_ventas['banda_sup'] = pd.cut(
    df_ventas['Sup_Cubierta_m2'],
    bins=[0, 35, 55, 75, 100, 999],
    labels=['<35', '35-55', '55-75', '75-100', '>100'],
    right=True
)

df_ventas = df_ventas.merge(
    tabla_puente[['Comuna', 'Tipologia', 'banda_sup',
                  'Renta_estimada_mensual_USD', 'Match_count']],
    on=['Comuna', 'Tipologia', 'banda_sup'],
    how='left'
)

def asignar_confianza(mc):
    if pd.isna(mc): return 'baja'
    elif mc >= 20:  return 'alta'
    elif mc >= 5:   return 'media'
    else:           return 'baja'

df_ventas['confianza_renta'] = df_ventas['Match_count'].apply(asignar_confianza)
df_ventas['Renta_m2_USD']    = df_ventas['Renta_estimada_mensual_USD'] / df_ventas['Sup_Cubierta_m2']

print(f"Ventas con renta asignada: {df_ventas['Renta_estimada_mensual_USD'].notna().sum():,} ({df_ventas['Renta_estimada_mensual_USD'].notna().mean()*100:.1f}%)")
print(f"confianza_renta: {df_ventas['confianza_renta'].value_counts().to_dict()}")
print(f"Renta_estimada_mensual_USD median: {df_ventas['Renta_estimada_mensual_USD'].median():,.0f}")
print(f"Renta_m2_USD               median: {df_ventas['Renta_m2_USD'].median():.2f}")

Ventas con renta asignada: 3,444 (89.0%)
confianza_renta: {'alta': 2051, 'media': 1104, 'baja': 713}
Renta_estimada_mensual_USD median: 608
Renta_m2_USD               median: 12.26


In [ ]:
# 2.11 — Celda de configuración
TASA_BONO_USD     = 0.10
MANTENIMIENTO_PCT = 0.03
VACANCIA_MESES    = 1
ABL_PCT_PRECIO    = 0.001

In [ ]:
# Fallback: tabla puente sin banda_sup para cubrir combinaciones sin match
tabla_puente_fallback = (
    df_alquileres.dropna(subset=['Comuna', 'Tipologia'])
    .groupby(['Comuna', 'Tipologia'], observed=True)
    .agg(
        Renta_estimada_mensual_USD=('Renta_largo_mensual_USD', 'median'),
        Match_count               =('Renta_largo_mensual_USD', 'count')
    )
    .reset_index()
)

# Aplicar solo a las 49 filas sin renta
mask_sin_renta = df_ventas['Renta_estimada_mensual_USD'].isna() & df_ventas['Comuna'].notna()

df_ventas.loc[mask_sin_renta, 'Renta_estimada_mensual_USD'] = (
    df_ventas[mask_sin_renta]
    .merge(tabla_puente_fallback, on=['Comuna', 'Tipologia'], how='left', suffixes=('', '_fb'))
    ['Renta_estimada_mensual_USD_fb'].values
)
df_ventas.loc[mask_sin_renta, 'Match_count'] = (
    df_ventas[mask_sin_renta]
    .merge(tabla_puente_fallback, on=['Comuna', 'Tipologia'], how='left', suffixes=('', '_fb'))
    ['Match_count_fb'].values
)

# Recalcular KPIs afectados
df_ventas['Renta_m2_USD'] = df_ventas['Renta_estimada_mensual_USD'] / df_ventas['Sup_Cubierta_m2']
df_ventas['confianza_renta'] = df_ventas['Match_count'].apply(asignar_confianza)

rent_anual = df_ventas['Renta_estimada_mensual_USD'] * 12
precio_aj  = df_ventas['Precio_ajustado_USD']
abl      = precio_aj * ABL_PCT_PRECIO
mant     = precio_aj * MANTENIMIENTO_PCT
vacancia = df_ventas['Renta_estimada_mensual_USD'] * VACANCIA_MESES

df_ventas['Rentabilidad_bruta_anual'] = rent_anual / precio_aj
df_ventas['Rentabilidad_neta_anual']  = (rent_anual - abl - mant - vacancia) / precio_aj
df_ventas['Payback_anios']            = precio_aj / rent_anual
df_ventas['Yield_gap_vs_bonos']       = df_ventas['Rentabilidad_bruta_anual'] - TASA_BONO_USD

print(f"Con Comuna pero sin renta tras fallback: {(df_ventas['Renta_estimada_mensual_USD'].isna() & df_ventas['Comuna'].notna()).sum()}")
print(f"Nulos Renta_estimada_mensual_USD: {df_ventas['Renta_estimada_mensual_USD'].isna().sum()}")

Con Comuna pero sin renta tras fallback: 0
Nulos Renta_estimada_mensual_USD: 375


####Parámetros de configuración

Se centralizan los supuestos de negocio que afectan los KPIs de retorno. Modificar cualquiera de estos valores recalcula automáticamente toda la cadena de indicadores, lo que permite al fondo realizar análisis de sensibilidad (por ejemplo, evaluar el impacto de una suba en la tasa de bonos sobre el yield gap, o de un aumento en los costos de mantenimiento sobre la rentabilidad neta).

####**KPIs de retorno**

Se calculan los indicadores centrales para la decisión de inversión del fondo. Rentabilidad_bruta_anual es el ratio de ingreso anual sobre precio ajustado, sin descontar gastos. Rentabilidad_neta_anual descuenta ABL (0,1% del precio), mantenimiento (3% anual) y un mes de vacancia, aproximando el retorno efectivo del inversor. Payback_anios es el tiempo de recupero de la inversión a renta constante. Yield_gap_vs_bonos mide el diferencial respecto a la tasa de bonos soberanos en USD, que es el costo de oportunidad relevante para un fondo que opera con capital propio en Argentina — si el gap es negativo, el activo no supera el rendimiento del bono y la inversión requiere justificación adicional por apreciación o diversificación.


In [ ]:
# 2.12 — KPIs de retorno
rent_anual = df_ventas['Renta_estimada_mensual_USD'] * 12
precio_aj  = df_ventas['Precio_ajustado_USD']

df_ventas['Rentabilidad_bruta_anual'] = rent_anual / precio_aj

abl      = precio_aj * ABL_PCT_PRECIO
mant     = precio_aj * MANTENIMIENTO_PCT
vacancia = df_ventas['Renta_estimada_mensual_USD'] * VACANCIA_MESES

df_ventas['Rentabilidad_neta_anual'] = (rent_anual - abl - mant - vacancia) / precio_aj
df_ventas['Payback_anios']           = precio_aj / rent_anual
df_ventas['Yield_gap_vs_bonos']      = df_ventas['Rentabilidad_bruta_anual'] - TASA_BONO_USD

print(f"Rentabilidad_bruta_anual median: {df_ventas['Rentabilidad_bruta_anual'].median():.4f}")
print(f"Rentabilidad_neta_anual  median: {df_ventas['Rentabilidad_neta_anual'].median():.4f}")
print(f"Payback_anios            median: {df_ventas['Payback_anios'].median():.1f}")
print(f"Yield_gap_vs_bonos       median: {df_ventas['Yield_gap_vs_bonos'].median():.4f}")

Rentabilidad_bruta_anual median: 0.0621
Rentabilidad_neta_anual  median: 0.0259
Payback_anios            median: 16.1
Yield_gap_vs_bonos       median: -0.0379


####**Liquidez de oferta por comuna**

Liquidez_oferta_comuna mide la cantidad de publicaciones activas por habitante en cada comuna. Para el fondo, que evalúa múltiples propiedades simultáneamente, la liquidez del mercado es una restricción operativa: una comuna con rentabilidad atractiva pero muy pocas unidades disponibles no es escalable. Este indicador permite filtrar comunas donde el volumen de oferta es insuficiente para desplegar capital a escala.


In [ ]:
# 2.13 — Liquidez_oferta_comuna
POB_COMUNAL = DELITOS_COMUNAS[['Comuna', 'poblacion']].copy()

publicaciones_ventas = (
    df_ventas.groupby('Comuna').size().reset_index(name='publicaciones')
)
liquidez = publicaciones_ventas.merge(POB_COMUNAL, on='Comuna', how='left')
liquidez['Liquidez_oferta_comuna'] = liquidez['publicaciones'] / liquidez['poblacion']

df_ventas     = df_ventas.merge(    liquidez[['Comuna', 'Liquidez_oferta_comuna']], on='Comuna', how='left')
df_alquileres = df_alquileres.merge(liquidez[['Comuna', 'Liquidez_oferta_comuna']], on='Comuna', how='left')

print(liquidez[['Comuna', 'publicaciones', 'Liquidez_oferta_comuna']].to_string(index=False))

 Comuna  publicaciones  Liquidez_oferta_comuna
    1.0            379                0.001841
    2.0            227                0.001437
    3.0            194                0.001034
    4.0             74                0.000339
    5.0            234                0.001307
    6.0            259                0.001471
    7.0            163                0.000739
    8.0             27                0.000144
    9.0             92                0.000569
   10.0            132                0.000795
   11.0            203                0.001069
   12.0            332                0.001659
   13.0            419                0.001811
   14.0            449                0.002035
   15.0            309                0.001692


####**Verificación final y guardado**

In [ ]:
# 2.14 — Verificación final y guardado
KPI_COLS = [
    'Precio_USD', 'Precio_ajustado_USD', 'Precio_m2_USD', 'Precio_m2_ajustado_USD',
    'Precio_m2_mediano_comuna', 'Renta_estimada_mensual_USD', 'Renta_m2_USD',
    'Match_count', 'Rentabilidad_bruta_anual', 'Rentabilidad_neta_anual',
    'Payback_anios', 'Yield_gap_vs_bonos', 'Distancia_subte_km', 'Distancia_parque_km',
    'Indice_amenities', 'Tasa_delitos_comuna', 'Volatilidad_Precio_m2_comuna',
    'Liquidez_oferta_comuna'
]

print("--- Nulos por columna KPI (ventas) ---")
nulos = df_ventas[KPI_COLS].isnull().sum()
print(nulos[nulos > 0].to_string() if (nulos > 0).any() else "Sin nulos en KPIs críticos")

print("\n--- Rangos plausibles (ventas) ---")
print(df_ventas[KPI_COLS].describe().loc[['min', '50%', 'max']].T.to_string())

print("\n--- Distribución Match_count ---")
print(df_ventas['Match_count'].value_counts(dropna=False).head(10).to_string())

print(f"\n--- confianza_renta ---")
print(df_ventas['confianza_renta'].value_counts().to_string())
print(f"Filas confianza_renta = baja: {(df_ventas['confianza_renta'] == 'baja').sum():,} ({(df_ventas['confianza_renta'] == 'baja').mean()*100:.1f}%)")

df_ventas.to_csv('dataset_ventas_features.csv', index=False)
df_alquileres.to_csv('dataset_alquileres_features.csv', index=False)

print(f"dataset_ventas_features.csv     — shape: {df_ventas.shape}")
print(f"dataset_alquileres_features.csv — shape: {df_alquileres.shape}")

--- Nulos por columna KPI (ventas) ---
Precio_m2_USD                     2
Precio_m2_ajustado_USD            2
Precio_m2_mediano_comuna        375
Renta_estimada_mensual_USD      375
Renta_m2_USD                    375
Match_count                     375
Rentabilidad_bruta_anual        375
Rentabilidad_neta_anual         375
Payback_anios                   375
Yield_gap_vs_bonos              375
Distancia_subte_km              375
Distancia_parque_km             375
Tasa_delitos_comuna             375
Volatilidad_Precio_m2_comuna    375
Liquidez_oferta_comuna          375

--- Rangos plausibles (ventas) ---
                                      min            50%           max
Precio_USD                    4500.000000  133250.000000  2.600000e+06
Precio_ajustado_USD           4279.050000  126707.425000  2.472340e+06
Precio_m2_USD                  331.858407    2472.222222  1.225000e+05
Precio_m2_ajustado_USD         315.564159    2350.836111  1.164852e+05
Precio_m2_mediano_comuna      